sklearn/utils/__init__.py

```diff
@@ -197,10 +204,14 @@ def _determine_key_type(key, accept_slice=True):
             raise ValueError(err_msg)
         return key_type.pop()
     if hasattr(key, "dtype"):
-        try:
-            return array_dtype_to_str[key.dtype.kind]
-        except KeyError:
-            raise ValueError(err_msg)
+        if hasattr(key.dtype, "kind"):
+            try:
+                return array_dtype_to_str[key.dtype.kind]
+            except KeyError:
+                raise ValueError(err_msg)
+        else:
+            # A `dtype` for a non-numpy library may not have a kind attribute
+            return "an-array-dtype"
```

Comment:
I think I prefer what is suggested by @betatim in https://github.com/scikit-learn/scikit-learn/pull/28407/files#diff-66316a90c42bbe375eef74f1b9dcbd6a1b5b99f60feb692f676067a633d30f60R305 instead.

Reply:
Agreed, that does look a bit cleaner indeed.

Reply:
Let's wait for the merge of #28407 first then.

Reply:
I aligned this PR with the code from #28407.